bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java

```diff
@@ -1300,29 +1308,68 @@ public void updateChannelDefinitions(Map<String, Channel> dynChannels) {
             return; // already done
         }
 
-        try {
-            // Get subset of those channels that currently do not exist
-            List<Channel> existingChannels = getThing().getChannels();
-            for (Channel channel : existingChannels) {
-                String id = channel.getUID().getId();
-                if (dynChannels.containsKey(id)) {
-                    dynChannels.remove(id);
+        List<Channel> channelMap = new ArrayList<>(this.getThing().getChannels());
+        List<Channel> removeChannels = new ArrayList<>();
+        List<Channel> upgradeChannels = new ArrayList<>();
+
+        // Step 1: Replace existing channels with dynamically created ones
+        for (Map.Entry<String, Channel> channel : dynChannels.entrySet()) {
+            String id = channel.getKey();
+            channelMap.removeIf(c -> (c.getUID().getId().equals(id)));
+            channelMap.add(channel.getValue());
+        }
+
+        // Step 2: check channel definition s for all channels to be up-to-date
+        for (Channel channel : channelMap) {
+            try {
+                boolean upgrade = false;
+                String channelId = channel.getUID().getId();
+                ChannelTypeUID uid = channel.getChannelTypeUID();
+                if (!ShellyChannelDefinitions.hasDefinition(channelId) || uid == null) {
+                    continue;
+                }
+                String typeId = uid.getBindingId().equals("system") ? uid.toString() : uid.getId();
+                ShellyChannel channelDef = ShellyChannelDefinitions.getDefinition(channelId);
+                ThingHandlerCallback callback = getCallback();
+                if (channelDef != null && getThing().getChannel(channel.getUID().getId()) != null && callback != null) {
+                    ChannelBuilder channelBuilder = callback.editChannel(thing, channel.getUID());
+                    if (!channelDef.typeId.equals(typeId)) {
+                        channelBuilder.withType(channelDef.getChannelTypeUID());
+                        upgrade = true;
+                    }
+                    String acceptedItemType = getString(channel.getAcceptedItemType());
+                    if (!acceptedItemType.isEmpty() && !channelDef.itemType.equals(acceptedItemType)) {
+                        channelBuilder.withAcceptedItemType(channelDef.itemType);
+                        upgrade = true;
+                    }
+                    if (upgrade) {
+                        logger.debug(
+                                "{}: Upgrade channel definition for channel {} (typeId {} -> {}, acceptedItemType {} -> {})",
+                                thingName, channelId, typeId, channelDef.typeId, acceptedItemType, channelDef.itemType);
+                        removeChannels.add(channel);
+                        upgradeChannels.add(channelBuilder.build());
+                    }
                 }
```

Comment:
Maybe some background: UOM changes in OH 4.0 brought up various mismatch between channel definitions and channel updates in the code. In addition the number of system defined channels were extended and I saw an issue that those should be used where applicable, so I decided to update various definitions to fix UOM issues or move to system defined types.

I looked into the upgrade mechanism, but that is not a feasible approach here, because the binding supports more than 50 different devices, some of them supporting multiple operation modes, single or multiple relays etc. The standard upgrade mechanism has no wildcard for the thing type, so I can't specify a new channel definition independent form the thing type. That means that I have to duplicate the upgrade statements 20 or more times to apply this to all thing types.

Therefore I created my own auto-upgrade, by matching the existing channel definitions to the binding internal ones- I check the typeId and acceptedItemType. If there is a mismatch I update the channel definition. On the other hand the channel might already exist and needs an update or is created newly depending on the current device setup (channels are created based on device model and configuration; if config is changed the binding auto-adds missing channels). Therefore I build a list of channels to be removed and one of that to be added. Overall I try to reduce the update to a minimum to prevent re-writing the channel information each time the binding initializes a thing (on start-up, suspend/resume or connection recovery). 

Maybe the code can be optimized, but it works. If might happen that the binding updates the channel type, e.g. for humidity This gets replaced by system:xxx and that works fine. However, when the channel type was ampere, but by mistake was declared as watt the binding updates the channel information, but from my understanding I can't update the item definition too. Therefore the user needs to remove the item linkage and re-add it again.

Open on other solutions, but I think moving to system defined types makes sense and auto-upgrade reduces the effort for the user, but can't fix any inconsistency.

Reply:
Perhaps @J-N-K has some input here?

Reply:
fyi: what I try to achive
- There are existing things, which have incorrect channel types vs. channel updates, which are creating UOM issues or in case of percentage just show 1 instead of 100% etc. The main objective is to fix those definitions when the new binding version initializes a thing. Those channels requiring or new channel type or have a mismatching acceptedItemType go into the upgradeChannel list. The upgrade mechanism implemented by the framework is not feasable, because I would need to duplicate the same channel information per thing rather than using * as thing type - define it once, use if for all thing types in the same way.
- A binding has existing channels, which should be kept on an upgrade. However, if a channel definition needs an update, it replaces the existing defiition, so I'm adding those to the list removeChannels.
- At the same time maybe new channels need to be added, e.g. the user change the device config and added a sensor, which then gets represented by new channels. Those channels go into the newChannels list.
- I want to minimize the overhead by re-writing all definitions, so I do the update at once
- And finally we have the topic iterating those list while deleting entries from the list at the same time. Therefore I create the lists while processing the channel and then do the update afterwards at once.

At least that's what I try to achieve, maybe there is a better way or this could be done more efficient. In fact the code above runs well (no issues anymore). The only remaining problem is that I can't change the item type, which means the user has to delete the current item and re-create the linkage so channel type and item type are in sync. Again, maybe there is a way even to fix this issue, but I don't know this option.

Reply:
Implementing updates within the binding was my first approach (back in ESH days), but the majority of maintainers (and I agree with that to a certain extend nowadays) thought that this is not the way to go, because it creates a lot of different code and makes especially more than one upgrade (i.e. version 1 -> version 2 -> version 3) extremely difficult. 

OTOH I see the issue you have with creating instructions for such a large number of things and channels. Do you think that you can catch all issues with your upgrade code or will there still be cases where a "remove thing/re-create thing" action is necessary? And if so, can you guess what percentage of users is affected?

Reply:
@J-N-K  As I described: There is issue remaining: The binding can't change the item type of linked to a channel with an updated channel type. Does the framework upgrade support this?

The UOM issues came up several times in the thread, but I don't know the percentage. On the other side every setup using percentage values (like humidity) or using the power meter channels (e. there was an issue with  Amp and Watt).

Can users resolve the issues by re-creating thing: yes, if they get the idea and trust OH, because OH restores the channel linkage. Nevertheless, I would expect that a mismatch in item type is also not resolved by this.

An easy solution would be extending the upgrade support to allow definitions for shelly:*, but I suppose this is not planned?